test/test-tty-mouse-tracking.c

```diff
@@ -0,0 +1,1440 @@
+  uv_run(loop, UV_RUN_DEFAULT);
+  ASSERT(assert_same(expected, actual));
+
+  button = 4 + 8 + 16;
+  expected->len = 0;
+  append_expected(expected, 0 + button + ' ', start.X + ' ', start.Y + ' ',
+                  'M', X10);
+  append_expected(expected, 3 + button + ' ', start.X + ' ', start.Y + ' ',
+                  'M', X10);
+  write_console_mouse_click(&tty_out, &tty_in, start,
+                            FROM_LEFT_1ST_BUTTON_PRESSED,
+                            (SHIFT_PRESSED |
+                             LEFT_ALT_PRESSED |
+                             LEFT_CTRL_PRESSED),
+                            0, BOTH);
+  uv_run(loop, UV_RUN_DEFAULT);
+  ASSERT(assert_same(expected, actual));
+
+  /* Test whether button number when button release is correctly generated. */
+  write_console(&tty_out, SET_ENC_SGR);
+  expected->len = 0;
+  append_expected(expected, 0, start.X, start.Y, 'M', SGR);
+  append_expected(expected, 1, start.X, start.Y, 'M', SGR);
+  append_expected(expected, number_of_buttons > 2 ? 2 : 1,
+                  start.X, start.Y, 'M', SGR);
+  append_expected(expected, 1, start.X, start.Y, 'm', SGR);
+  append_expected(expected, number_of_buttons > 2 ? 2 : 1,
+                  start.X, start.Y, 'm', SGR);
+  append_expected(expected, 0, start.X, start.Y, 'm', SGR);
+  write_console_mouse_click(&tty_out, &tty_in, start,
+                            FROM_LEFT_1ST_BUTTON_PRESSED, 0, 0,
+                            PRESS);
+  write_console_mouse_click(&tty_out, &tty_in, start,
+                            FROM_LEFT_1ST_BUTTON_PRESSED |
+                            FROM_LEFT_2ND_BUTTON_PRESSED, 0, 0, PRESS);
+  write_console_mouse_click(&tty_out, &tty_in, start,
+                            FROM_LEFT_1ST_BUTTON_PRESSED |
+                            FROM_LEFT_2ND_BUTTON_PRESSED |
+                            RIGHTMOST_BUTTON_PRESSED, 0, 0, PRESS);
+  /* Release 2nd button from the left. */
+  write_console_mouse_click(&tty_out, &tty_in, start,
+                            FROM_LEFT_1ST_BUTTON_PRESSED |
+                            RIGHTMOST_BUTTON_PRESSED, 0, 0, PRESS);
+  /* Release rightmost button. */
+  write_console_mouse_click(&tty_out, &tty_in, start,
+                            FROM_LEFT_1ST_BUTTON_PRESSED, 0, 0, PRESS);
+  write_console_mouse_click(&tty_out, &tty_in, start, 0, 0, 0, RELEASE);
+  uv_run(loop, UV_RUN_DEFAULT);
+  ASSERT(assert_same(expected, actual));
+
+
+  finalize_buf(expected);
+  finalize_buf(actual);
+  MAKE_VALGRIND_HAPPY();
+  return 0;
+}
+
+TEST_IMPL(tty_mouse_tracking_mode_x10) {
+  uv_tty_t tty_in, tty_out;
+  uv_loop_t* loop;
+  uv_buf_t  *expected, *actual;
+  INPUT_RECORD key_records[2];
+  COORD start = {1, 1}, end = {4, 4}, current_pos;
+  DWORD written;
+  int resolution;
+
+  loop = uv_default_loop();
+
+  initialize_tty(&tty_in, &tty_out);
+  initialize_buf(&expected);
+  initialize_buf(&actual);
+
+  uv_handle_set_data((uv_handle_t*)&tty_in, (void*)actual);
```

Comment:
```suggestion
  uv_handle_set_data((uv_handle_t*) &tty_in, actual);
```
or (more typical):
```suggestion
  tty_in->data = actual;
```

Reply:
It is rarely used in practice, but I think it should be used if setters and getters are defined.